spec/bucket_store_integration_spec.rb

```diff
@@ -24,55 +26,129 @@
   end
 
   shared_examples "adapter integration" do |base_bucket_uri|
-    # This is presented as a single idempotent test as otherwise resetting state between execution
-    # makes things very complicated with no huge benefits.
-
-    it "has a consistent interface" do
-      # Write 201 files
-      file_list = []
-      201.times do |i|
-        filename = "file#{(i + 1).to_s.rjust(3, '0')}.txt"
-        file_list << filename
-
-        # the body of the file is the filename itself
-        described_class.for("#{base_bucket_uri}/prefix/#{filename}").upload!(filename)
+    context "using #{base_bucket_uri}" do
+      before do
+        described_class.for(base_bucket_uri).list.each do |path|
+          described_class.for(path).delete!
+        end
       end
 
-      # Add some files with spaces
-      described_class.for("#{base_bucket_uri}/prefix/i have a space.txt").
-        upload!("i have a space.txt")
-      described_class.for("#{base_bucket_uri}/prefix/another space.txt").
-        upload!("another space.txt")
-
-      file_list << "i have a space.txt"
-      file_list << "another space.txt"
-
-      # List with prefix should only return the matching files
-      expect(described_class.for("#{base_bucket_uri}/prefix/file1").list.to_a.size).to eq(100)
-      expect(described_class.for("#{base_bucket_uri}/prefix/file2").list.to_a.size).to eq(2)
-      expect(described_class.for("#{base_bucket_uri}/prefix/").list.to_a.size).to eq(203)
-
-      # List (without prefixes) should return everything
-      expect(described_class.for(base_bucket_uri.to_s).list.to_a).
-        to match_array(file_list.map { |filename| "#{base_bucket_uri}/prefix/#{filename}" })
-
-      # We know the content of the file, we can check `.download` returns it as expected
-      all_files = file_list.map do |filename|
-        [filename, "#{base_bucket_uri}/prefix/#{filename}"]
-      end
-      all_files.each do |content, key|
-        expect(described_class.for(key).download[:content]).to eq(content)
+      context "when handling streaming uploads / downloads" do
+        after do
+          file_to_read.close
+          file_to_write.close
+        end
+
+        shared_examples "consistently handles different file modes" do
```

Comment:
I'm a bit wary of `shared_examples` within another shared example as it's starting to be hard to follow what's going on. Wonder if we can simplify these tests a bit, and split them out somehow? With the delete-all in the `before` block, I reckon we can isolate things a bit more now instead of having a single massive test

Reply:
Re-worked it a bit and made binary files somewhat of a special case e.g. listing files / deleting shouldn't care if the files stored in GCS are binary or textual